chore: create fixture for Fuel node

deekerno opened the pull request:
## Changelog
- Add fixture for Fuel node and adjusts E2E tests to use it
- Remove `fuel-node` package from workspace
- Remove `fuel-node` build step from CI
- Remove `e2e.bash` test script
- Adjust E2E tests to remove unnecessary asserts
- Add small `hello-world-test-node` package that wraps the aforementioned test fixture
- Adjusts documentation for Hello World example

## Testing Steps
CI should pass with new Fuel node fixture. Additionally, the "Hello World" example has been adjusted; run through both the example README and the example in the documentation, and verify that everything works.

## Notes
- No big improvement in CI runtime; however, it does allow for all of our testing to be done through `cargo` actions and for us not to use extra scripts.
- A number of tests needed to be adjusted to either remove asserts that weren't properly testing values or to delete records from tables at test invocation to ensure a clean slate.

ra0x3 commented on packages/fuel-indexer-tests/tests/e2e/graphql_api_sqlite.rs:
```diff
@@ -17,92 +17,96 @@ use serde_json::{Number, Value};
-    assert_eq!(v[0]["int_optional_some"], Value::from(Number::from(999)));
-    assert_eq!(v[0]["addr_optional_none"], Value::from(None::<&str>));
-}
+// #[actix_web::test]
```
I know we've done this in the past, and this is temporary (not running these tests), but let's remove these fully until they're added back (should be easy to just copy the test from an old commit).

ra0x3 commented on Cargo.toml:
```diff
@@ -17,7 +17,6 @@ members = [
    "packages/fuel-indexer-plugin",
    "packages/fuel-indexer-schema",
    "packages/fuel-indexer-tests",
-   "packages/fuel-indexer-tests/components/fuel-node",
```
🙌

ra0x3 commented on .github/workflows/ci.yml:
```diff
@@ -253,25 +252,17 @@ jobs:
-            --chain-config packages/fuel-indexer-tests/assets/test-chain-config.json \
-            --contract-bin packages/fuel-indexer-tests/contracts/fuel-indexer-test/out/debug/fuel-indexer-test.bin &
+          command: test
+          args: --features e2e,postgres e2e -- --test-threads=1
```
```suggestion
          args: --features e2e,postgres e2e --locked -- --test-threads=1
```
- I remember specifically that [`serial`](https://docs.rs/serial_test/latest/serial_test/) and `--test-threads` weren't working for me I was testing DB isolation
- Looks like some change that was made allows this now - love it

ra0x3 commented on .github/workflows/ci.yml:
```diff
@@ -253,25 +252,17 @@ jobs:
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --features e2e,sqlite e2e -- --test-threads=1
```
```suggestion
          args: --features e2e,sqlite e2e  --locked -- --test-threads=1
```

ra0x3 reviewed (changes requested):
@deekerno Looks mostly OK, thanks for tackling this (it's long overdue). 

- Left some minor feedback
- ✅ Tests pass for me locally as well
- Remember we still use the `fuel-node` in our `hello-world` example. So if you replace it (which is ok, and preferred) we will need a new way to:
  - Start `fuel-core`
  - Deploy a contract
  - Interact with that contract
  - I know fuel-core allows `fuel-core --chain test-chain-config.json` but I'm not sure how to deploy the contract and such - might be a good time to learn.
    - We would need to update the docs for that example accordingly
  - We also won't be needing the `Dockerfile.fuel-node` any more either

deekerno commented:
@ra0x3 I've addressed your feedback. By running `fuel-core run --chain-config [path_to_test_chain_config]` and deploying the assets to the service, you should be able to run the bin for `hello-world-data` and it should deploy the contract and generate the events.

ra0x3 commented on docs/src/examples/hello-world.md:
````diff
@@ -9,13 +9,13 @@ A "Hello World" type of program for the Fuel Indexer service.
 //!
 //! ```bash
-//! cargo run --bin fuel-node
+//! fuel-core run --chain packages/fuel-indexer-tests/assets/test-chain-config.json
````
@deekerno  
- So where is the contract deployment here? 
- The `fuel-node` that is being removed uses a deployed contract with a funded wallet that we can make contract calls against
- If we're gonna fully remove the `fuel-node` we need to update the example with these replacement steps
- I believe this just starts the node?

ra0x3 commented on docs/src/examples/hello-world.md:
````diff
@@ -28,7 +28,7 @@ A "Hello World" type of program for the Fuel Indexer service.
 //!
 //! ```bash
-//! cargo run --bin hello-bin
+//! cargo run --bin hello-world-data
````
Oh nice catch 👌

ra0x3 commented on docs/src/examples/hello-world.md:
````diff
@@ -9,13 +9,13 @@ A "Hello World" type of program for the Fuel Indexer service.
 //!
 //! ```bash
-//! cargo build -p hello-index --release --target wasm32-unknown-unknown
+//! cargo build -p hello_index --release --target wasm32-unknown-unknown
````
🙏

ra0x3 commented on examples/hello-world/hello-index/hello_index.manifest.yaml:
```diff
@@ -4,7 +4,7 @@ abi: examples/hello-world/contracts/greeting/out/debug/greeting-abi.json
 graphql_schema: examples/hello-world/schema/hello_index.schema.graphql
 module:
-  wasm: target/wasm32-unknown-unknown/debug/hello_index.wasm
+  wasm: target/wasm32-unknown-unknown/release/hello_index.wasm
```
**dj khaled voice** "Anotha one"

ra0x3 reviewed (changes requested):
@deekerno 
- Looks good
- However, the `hello-world` example updates only show how to start a new `fuel-node` with a chain config? (unless I'm missing something)
- We would need the example to show how to deploy the contract as well
- I realize this might seem out of scope but the `fuel-node` that is being removed did all of that 
- The `docker-compose` from the `hello-world` example would also have to be updated - as it's referencing a removed `Dockerfile.fuel-node`

deekerno commented:
@ra0x3 I've adjusted the fixture one more time. I made it so that we can choose what contract bin path to pass in; using this, I created a small test node package for the hello world example in which all it does is spawn the test fixture.

ra0x3 commented on packages/fuel-indexer-tests/tests/e2e/graphql_api_postgres.rs:
```diff
@@ -64,6 +73,10 @@ async fn test_can_return_query_response_with_all_fields_required_postgres() {
 #[actix_web::test]
 #[cfg(all(feature = "e2e", feature = "postgres"))]
 async fn test_can_return_query_response_with_nullable_fields_postgres() {
+    let fuel_node_handle = tokio::spawn(setup_test_fuel_node(
```
I see pretty much every test that uses `setup_test_fuel_node` has this same snippet:


```rust
    let fuel_node_handle = tokio::spawn(setup_test_fuel_node(
        get_test_chain_config_path(),
        Some(get_test_contract_bin_path()),
    ));
```
Is there any way we can cut down on this boilerplate if we know that all tests will be using the same params? E.g.,:

```rust
    let fuel_node_handle = tokio::spawn(setup_test_fuel_node());
```
And if some non-test code is calling `setup_test_fuel_node` (e.g., `hello-world-test-node`) maybe we can wrap `setup_test_fuel_node` in something like `setup_example_fuel_node` which would call  `setup_test_fuel_node` under the hood?

deekerno replied:
Yeah, using your suggestion in the last sentence is what I'll do.

ra0x3 commented on ci/Dockerfile.fuel-node:
```diff
@@ -31,7 +31,7 @@ COPY --from=planner /build/recipe.json recipe.json
 ENV SQLX_OFFLINE=true
```
- Everywhere in the code, the test node for the `hello-world` example is referenced as `hello-world-test-node`. 
- Could we update this filename to reflect that? It seems to still have the old/generic `fuel-node` name.

ra0x3 commented on .github/workflows/ci.yml:
```diff
@@ -251,25 +250,11 @@ jobs:
-      - name: Stop testing components
-        run: kill -9 $(lsof -ti:4000)
+          command: test
+          args: --features e2e,postgres e2e --locked -- --test-threads=1
```
👌

ra0x3 reviewed (changes requested):
Getting the following when running the hello-world example per the updated README

```
 forc index deploy \ 
   --path hello-index \
   --output-dir-root /Users/rashad/development/repos/fuel-indexer \
   --url http://0.0.0.0:29987 \
   --target wasm32-unknown-unknown \
   --release
▹▹▹▹▹ ⏰ Building...                                                                                                      Finished release [optimized] target(s) in 0.73s
▪▪▪▪▪ ✅ Build succeeded.                                                                                             Deploying indexer at /Users/rashad/development/repos/fuel-indexer/examples/hello-world/hello-index/hello_index.manifest.yaml to http://0.0.0.0:29987/api/index/fuel_examples/hello_index
▸▹▹▹▹ 🚀 Deploying...                                                                                                 thread 'main' panicked at 'Failed to deploy indexer.: reqwest::Error { kind: Body, url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(0.0.0.0)), port: Some(29987), path: "/api/index/fuel_examples/hello_index", query: None, fragment: None }, source: hyper::Error(ChannelClosed) }', plugins/forc-index/src/ops/forc_index_deploy.rs:103:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[1]    12520 abort      forc index deploy --path hello-index --output-dir-root  --url  --target
```

I don't believe the new README mentions Docker?

ra0x3 commented on packages/fuel-indexer-tests/tests/e2e/indexing_postgres.rs:
```diff
@@ -24,6 +28,8 @@ use tokio::time::{sleep, Duration};
 #[cfg(all(feature = "e2e", feature = "postgres"))]
 async fn test_can_trigger_and_index_events_with_multiple_args_in_index_handler_postgres()
 {
+    let fuel_node_handle = tokio::spawn(setup_example_test_fuel_node());
```
:chef's kiss:

ra0x3 reviewed (changes requested):
Hm, getting this error after a successful deploy (I made the local change to update the Dockerfile name just FYI)

```text
n: Name or service not known
hello-world-fuel-indexer-1  | Feb 03 17:57:38.259  INFO sqlx::query: BEGIN; rows affected: 0, rows returned: 0, elapsed: 242.912µs
hello-world-fuel-indexer-1  | Feb 03 17:57:38.259  INFO sqlx::query: COMMIT; rows affected: 0, rows returned: 0, elapsed: 253.683µs
hello-world-fuel-indexer-1  | Feb 03 17:57:38.259  INFO fuel_indexer::executor: No new blocks to process, sleeping.
hello-world-fuel-indexer-1  | Feb 03 17:57:39.266 ERROR fuel_indexer::executor: Failed to retrieve blocks: Error making HTTP request: error sending request for url (http://fuel-node:4000/graphql): error trying to connect: dns error: failed to lookup address information: Name or service not known
hello-world-fuel-indexer-1  | Feb 03 17:57:39.267  INFO sqlx::query: BEGIN; rows affected: 0, rows returned: 0, elapsed: 257.298µs
hello-world-fuel-indexer-1  | Feb 03 17:57:39.268  INFO sqlx::query: COMMIT; rows affected: 0, rows returned: 0, elapsed: 192.477µs
hello-world-fuel-indexer-1  | Feb 03 17:57:39.268  INFO fuel_indexer::executor: No new blocks to process, sleeping.
hello-world-fuel-indexer-1  | Feb 03 17:57:40.275 ERROR fuel_indexer::executor: Failed to retrieve blocks: Error making HTTP request: error sending request for url (http://fuel-node:4000/graphql): error trying to connect: dns error: failed to lookup address information: Name or service not known
hello-world-fuel-indexer-1  | Feb 03 17:57:40.275  INFO sqlx::query: BEGIN; rows affected: 0, rows returned: 0, elapsed: 208.325µs
hello-world-fuel-indexer-1  | Feb 03 17:57:40.276  INFO sqlx::query: COMMIT; rows affected: 0, rows returned: 0, elapsed: 213.295µs
hello-world-fuel-indexer-1  | Feb 03 17:57:40.276  INFO fuel_indexer::executor: No new blocks to process, sleeping.
hello-world-fuel-indexer-1  | Feb 03 17:57:41.285 ERROR fuel_indexer::executor: Failed to retrieve blocks: Error making HTTP request: error sending request for url (http://fuel-node:4000/graphql): error trying to connect: dns error: failed to lookup address information: Name or service not known
hello-world-fuel-indexer-1  | Feb 03 17:57:41.285  INFO sqlx::query: BEGIN; rows affected: 0, rows returned: 0, elapsed: 235.124µs
hello-world-fuel-indexer-1  | Feb 03 17:57:41.286  INFO sqlx::query: COMMIT; rows affected: 0, rows returned: 0, elapsed: 191.081µs
hello-world-fuel-indexer-1  | Feb 03 17:57:41.286  INFO fuel_indexer::executor: No new blocks to process, sleeping.
hello-world-fuel-indexer-1  | Feb 03 17:57:42.293 ERROR fuel_indexer::executor: Failed to retrieve blocks: Error making HTTP request: error sending request for url (http://fuel-node:4000/graphql): error trying to connect: dns error: failed to lookup address information: Name or service not known
```

deekerno commented:
Pausing work on this while we do the upgrade to `fuels-rs v0.35.0`.
